sidecar_acceptor.go

```diff
@@ -131,3 +193,268 @@ func (a *SidecarAcceptor) RegisterSidecar(ctx context.Context,
 
 	return nil
 }
+
+// ExpectChannel informs the acceptor that a new bid order was submitted for the
+// given sidecar ticket. We subscribe to auction events using the multisig key
+// we gave out when we registered the ticket.
+func (a *SidecarAcceptor) ExpectChannel(ctx context.Context,
+	t *sidecar.Ticket) error {
+
+	a.Lock()
+	defer a.Unlock()
+
+	if t.Order == nil {
+		return fmt.Errorf("order in sidecar ticket is missing")
+	}
+
+	nonce := t.Order.BidNonce
+	_, ok := a.pendingSidecarOrders[nonce]
+	if ok {
+		return fmt.Errorf("sidecar with order nonce %x is already "+
+			"registered", nonce[:])
+	}
+
+	// We didn't know about this ticket for this nonce before so let's now
+	// update its state in the database and start expecting a channel for it
+	// now.
+	t.State = sidecar.StateExpectingChannel
+	if err := a.store.UpdateSidecar(t); err != nil {
+		return fmt.Errorf("error updating sidecar: %v", err)
+	}
+
+	a.pendingSidecarOrders[nonce] = t
+
+	return a.client.SubscribeAccountUpdates(ctx, &keychain.KeyDescriptor{
+		KeyLocator: keychain.KeyLocator{
+			Family: keychain.KeyFamilyMultiSig,
+			Index:  t.Recipient.MultiSigKeyIndex,
+		},
+		PubKey: t.Recipient.MultiSigPubKey,
+	})
+}
+
+// HasPendingBatch returns true if the acceptor is currently working on
+// accepting a channel from a pending batch.
+func (a *SidecarAcceptor) HasPendingBatch() bool {
+	a.Lock()
+	defer a.Unlock()
+
+	return a.pendingBatch != nil
+}
+
+// matchPrepare handles an incoming OrderMatchPrepare message from the server.
+// Since we're only on the receiving end of a sidecar channel (which is always
+// a bid order) the tasks are simplified compared to normal bid order execution.
+func (a *SidecarAcceptor) matchPrepare(
+	msg *auctioneerrpc.OrderMatchPrepare) error {
+
+	// Parse and formally validate what we got from the server.
+	log.Tracef("Received prepare msg from server, batch_id=%x: %v",
+		msg.BatchId, spew.Sdump(msg))
+	batch, err := order.ParseRPCBatch(msg)
```

Comment:
Seems like a bit more cod here can be consolidated with the main (full client) flow for processing a batch. Don't consider it non-blocking, as of right now, still need to get through the rest of the diff.

Reply:
I wanted to keep this separate so the custom flow can be understood more easily instead of adding even more cases to the code in the normal RPC server.

Reply:
Gotcha yeah great decision IMO, albeit there's a _bit_ of duplication here, the overall logic is much easier to follow.